### ISSUE TITLE
Ayuda TG2 - Normalización y métricas

### ISSUE THREAD

**Issue by Mdelaf** [Exploratorio-DCC-PUC/Syllabus-2018-2#80]
A partir de las preguntas que surgieron en ayudantía hemos decido crear esta _issue_ para aclarar algunos conceptos y guiarlos en su tarea.

### Normalización

En palabras muy simples, normalizar los datos significa dividir cada número por un número más grande, de tal modo que todos los datos queden en el intervalo [0, 1]. 

Por ejemplo, si yo tengo una lista con los números `[0, 3, 4, 5, 8, 10]` yo puedo dividir todos los números por 10 y así los valores normalizados quedarían como `[0, 0.3, 0.4, 0.5, 0.8, 1]`. La regla de normalización más sencilla consiste en dividir por el valor máximo como acabamos de hacer en el ejemplo. Otra regla comúnmente usada es la norma `l2` que en vez de tomar el máximo de los valores, considera la lista de números como un vector y luego divide cada valor por el módulo del vector. 

Ahora bien, en el dataset de su tarea se pide que normalicen todos los atributos. Esto significa tomar cada columna del dataset por separado y normalizarla según sus valores. Sin duda es algo que pueden hacer en "Python plano" ahora que entienden el concepto, pero si deciden usar la función [`normalize` de `scikit-learn`](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html#sklearn.preprocessing.normalize) entonces consideren que uno le puede decir el tipo de norma (`max`, `l2` o `l1`) y lo más importante es especificar qué eje quieren normalizar. Si quieren normalizar las columnas entonces deben pasarle el argumento `axis=0`, **si no hacen esto van a estar normalizando las filas y eso no es correcto**. 

Recuerden además que solo deben normalizar los atributos de sus datos. No tiene sentido que normalicen el `id` ni el target (`calidad`). 

### Métricas para medir rendimiento 

Como vimos en la ayudantía y en en los ejemplos de la semana pasada, el `score` es una métrica que nos dice qué porcentaje de los ejemplos se clasifican correctamente. Si para _testear_ le entregamos 100 ejemplos y solo clasifica bien 20 de ellos, el score es `0.2`.

El `score` es una métrica muy útil y usada en clasificación, pero en algunos problemas puede no ser la más indicada. Piensen en su tarea, cuando el clasificador predice que un vino tiene calidad 8 y en realidad tenía calidad 9, para la métrica `score` eso es un error en la clasificación, al igual que si hubiera predicho calidad 3. Sin embargo el 8 está muy cerca del 9, por lo tanto la predicción no está tan mal. 

Es decir, `score` no considera si existe una "relación de cercanía" entre las clases. Para cada muestra el  `score` la cataloga como correcta o incorrecta, sin importarle la cercanía. Es por esta razón que probablemente el `score` que obtengan en sus clasificadores no sea muy alto.

Como se comentó en la ayudantía se pueden usar funciones de pérdida que tal vez sean mejores indicadores en el problema de la tarea. Scikit-learn incluye varias funciones de pérdida, pero podría no ser tan claro qué es lo que representan sus valores. Un ejemplo simple que podrían hacer a mano es calcular para cada dato :
`diff = |valor_real - valor_predicho|`

Y luego calcular el promedio de las diferencias. Esta función de pérdida les diría qué tanto se aleja su clasificador de la calidad real del vino (en promedio). 

Por último, recuerden que para comparar los clasificadores deben **aplicar las métricas sobre el set de test**. Si comparan con las métricas sobre el set de training solo estarán midiendo su capacidad para "recordar" los ejemplos que ya vieron. 

#### Comentarios adicionales

Como se dijo hoy, la librería `scikit-learn` permite hacer casi todas las cosas de la tarea con una sola línea. Sin embargo no van a obtener buenos resultados si no entienden bien lo que están haciendo, si no leen la documentación y si no ajustan los parámetros de cada función que llaman.

Por favor hagan buen uso de la documentación y cualquier duda la pueden preguntar acá en las _issues_.

Éxito con su tarea!
